WebKit::compileAndApplySandboxSlowCase can crash when logging an error

jelee53 opened the pull request:
#### d5fd1f9f14bc7448fab536d3dab984c6c04c5274
<pre>
WebKit::compileAndApplySandboxSlowCase can crash when logging an error
<a href="https://bugs.webkit.org/show_bug.cgi?id=297879">https://bugs.webkit.org/show_bug.cgi?id=297879</a>
<a href="https://rdar.apple.com/146431897">rdar://146431897</a>

Reviewed by Per Arne Vollan and Ben Nham.

We are receiving reports that WebKit::compileAndApplySandboxSlowCase can crash in rare cases.

* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::compileAndApplySandboxSlowCase):

Canonical link: <a href="https://commits.webkit.org/299175@main">https://commits.webkit.org/299175@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a83b675f61e4502df0360d4fe153d03cebc3f959

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118128 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37806 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28442 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124283 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70167 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2518c71f-d269-4103-b587-6e7e2e7ae62b) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38499 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46388 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89640 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59272 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/fb1d2501-388c-41db-9ddc-b77d79ed1617) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121080 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30664 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105917 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70133 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/5e6300fe-9041-4ea0-8464-1cd710591d3d) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29733 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67950 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100090 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24211 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127360 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45031 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33939 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98319 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45392 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102139 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98105 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43502 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21494 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/41476 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18824 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44903 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50577 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44363 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47708 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46052 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
```diff
@@ -552,8 +552,12 @@ static bool compileAndApplySandboxSlowCase(const String& profileOrProfilePath, b
 ALLOW_DEPRECATED_DECLARATIONS_END
         WTFLogAlways("%s: Could not initialize sandbox profile [%s], error '%s'\n", getprogname(), temp.data(), errorBuf);
-        for (size_t i = 0, count = parameters.count(); i != count; ++i)
-            WTFLogAlways("%s=%s\n", parameters.name(i), parameters.value(i));
```
Isn't the cause of the crash that we're passing an ASCIILiteral as the second argument to `WTFLogAlways` rather than a `char *`? Also, how did this even compile in the first place...

bnham replied:
I think you need something like:

```
WTFLogAlways("%s=%s\n", parameters.name(i).characters(), parameters.value(i));
```

`%s` should handle printing a null pointer just fine, it should output it like `(null)` or something like that.

mwyrzykowski replied:
I've found %s or %@ are not type checked in the format string against the arguments. I've encountered many local crashes during debugging when I accidentally print an NSString with %s or a String.utf8().data() with %@.

bnham reviewed (approved):
Right, so what was happening is that we were dying on e.g. a segfault at 0x10.

This is because `WTF::ASCIILiteral` is basically a `std::span`, and we're passing it in to a varargs function associated with the format specifier `%s`.

A `std::span` is a 16 byte struct that contains a pointer and a size. For a varargs function on Apple Silicon, a 16 byte struct gets passed via two registers.

So the printf function looks at the first arg to handle the first `%s`, which will be a pointer to a C string (associated with ASCIILiteral's first field, which is a pointer). This works fine. Then it looks at the second arg to handle the second `%s`, which it also expects to a pointer (associated with `paramters.value(i)` in this case, which is supposed to be a `const char *`). However, instead, the second argument is instead ASCIILiteral's second field, which is the length of the literal.

If the length of the literal is 16 (0x10), then you end up segfaulting at 0x10 like we see in the crash logs.

The second issue here is why `-Wformat` doesn't seem to be enabled on .mm files, as this snippet of code doesn't compile (as expected) in .cpp files.

jelee53 committed:
WebKit::compileAndApplySandboxSlowCase can crash when logging an error
https://bugs.webkit.org/show_bug.cgi?id=297879
rdar://146431897

Reviewed by Per Arne Vollan and Ben Nham.

We are receiving reports that WebKit::compileAndApplySandboxSlowCase can crash in rare cases.

* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::compileAndApplySandboxSlowCase):

Canonical link: https://commits.webkit.org/299175@main
